### PR TITLE
clang-tidy fixes

### PR DESCRIPTION
**Pull request by neheb**

### CODE REVIEW

**Review by mbunkus — CHANGES_REQUESTED** [review]
Thanks. All tests in the MKVToolNix test suite fail. A simple `mkvmerge -o out.mka in.mp3` segfaults.

Please investigate why this happens. Thanks.

**Comment by neheb** [comment]
Most likely clang-tidy: make member functions const

**Comment by neheb** [comment]
@mbunkus removed offending commit. I expect tests to pass now.

**Comment by mbunkus** [comment]
Thanks. From a cursory look the changes seem OK to me, but I still need to test them.

One thing I'm curious about: some variable declarations are converted to `auto` (which I like), while a lot of others that have been touched haven't been — even though it would make so much sense for them, too! For example:

```c++
    const KaxInternalBlock& MyBlock = *static_cast<KaxBlock *>(this->FindElt(EBML_INFO(KaxBlock)));
```

Why not for all of those, too?

**Comment by neheb** [comment]
@mbunkus So that clang-tidy check's fix routing adds const to the east. I had to manually fix it up. The autos were manually added as I thought they looked cleaner. I'll look into the others.

**Comment by neheb** [comment]
@mbunkus fixed.

**Comment by mbunkus** [comment]
Thanks. I'll look some more at it tomorrow.

One thing I'd like you to change already: all those changes to the `enums` in `KaxSemantic.h`. That file is auto-generated from the Matroska specs XML file with tools outside this repo. I'll adjust those to make those changes instead. Thanks.

**Comment by neheb** [comment]
Added an extra const commit.

**Comment by neheb** [comment]
Removed typedef commit.

**Comment by neheb** [comment]
out of curiosity, why does the code use *static_cast<X*>(y) instead of X(y) ?

**Comment by neheb** [comment]
@robUx4 know something about the last commit?

**Comment by robUx4** [comment]
> out of curiosity, why does the code use
> 
> ```
>  *static_cast<X*>(&y)
> ```
> 
> instead of X(y) ? That smells like an old MSVC compiler bug.

That would be to avoid using pointers further down. I'm not against this change but doesn't improve the code much, nor is related to clang-tidy fixes (or it's not mentioned in the commit).

**Comment by neheb** [comment]
it's not prefixed with clang-tidy, no. It was something I noticed after handling @mbunkus 's comments.

My thinking was it was some bugfix similar to https://github.com/Exiv2/exiv2/commit/3662c29f16

IIRC, you use old compilers.

**Comment by mbunkus** [comment]
Just tested the current commits: they fail miserably with memory errors all over (access out of bounds etc.). Will investigate, though probably not before the next release of libMatroska. Please don't merge for the time being.

File: src/KaxBlock.cpp [hunk 1]
```diff
@@ -856,8 +856,7 @@ void KaxBlockGroup::ReleaseFrames()
 void KaxInternalBlock::ReleaseFrames()
 {
   // free the allocated Frames
-  int i;
-  for (i=myBuffers.size()-1; i>=0; i--) {
+  for (size_t i=myBuffers.size()-1; i>=0; i--) {
```
Comment (mbunkus) [inline src/KaxBlock.cpp:857]: This commit, or rather this change, is responsible for some of the crashes. And it's rather easy to see why: `myBuffers` can be empty, meaning that `myBuffers.size()-1` will lead to an integer underflow with an unsigned `i`, leading to invalid memory accesses.

Additionally, you really cannot use `i>=0` as the check with unsigned variables. It'll always leads to an integer underflow, 'cause when `i==0` the loop will be run as `i==0` obviously satisfies `i>=0`, and afterwards `i--` will underflow, wrapping back around to `std::numeric_limits<size_t>::max()`, meaning it'll run as an endless loop, accessing out-of-bounds memory.

This whole chunk is just wrong. Either do an explicit cast to `int`, which would be fine, or use this:

```c++
  for (size_t i = myBuffers.size(); i > 0; i--) {
    // now always access as [i-1]
  }
```

I hate both variants, but it's what we got in C & C++, and as it's less code to fix & reason about, I prefer:

```c++
  for (int i=static_cast<int>(myBuffers.size())-1; i>=0; i--) {
```

with the rest unchanged. You don't even have to check for `.isEmpty()`.

File: src/KaxCues.cpp [hunk 2]
```diff
@@ -67,8 +67,8 @@ bool KaxCues::AddBlockGroup(const KaxBlockGroup & BlockRef)
 bool KaxCues::AddBlockBlob(const KaxBlockBlob & BlockReference)
 {
   // Do not add the element if it's already present.
-  bool present = std::any_of(myTempReferences.begin(), myTempReferences.end(),
-    [&](const KaxBlockBlob *myTempReference) { return myTempReference == &BlockReference; });
+  auto pr = myTempReferences;
```
Comment (mbunkus) [inline src/KaxCues.cpp:70]: This creates a _copy_ of `myTempReferences`. In this place this isn't harmful, but it does cost some performance unnecessarily. Please use a reference instead.

File: src/KaxCues.cpp [hunk 3]
```diff
@@ -80,10 +80,10 @@ bool KaxCues::AddBlockBlob(const KaxBlockBlob & BlockReference)
 void KaxCues::PositionSet(const KaxBlockBlob & BlockReference)
 {
   // look for the element in the temporary references
-  auto it = std::find_if(myTempReferences.begin(), myTempReferences.end(),
-    [&](const KaxBlockBlob *myTempReference){ return myTempReference == &BlockReference; });
+  auto pr = myTempReferences;
```
Comment (mbunkus) [inline src/KaxCues.cpp:83]: This creates a _copy_ of `myTempReferences`. In this place this _is_ actually harmful, as the iterator is later used in `myTempReference.erase(it)` — but after your change the `it` iterator belongs to a different object. Again, just use a reference instead.

File: src/KaxBlock.cpp [hunk 4]
```diff
@@ -867,8 +867,8 @@ void KaxInternalBlock::ReleaseFrames()
 void KaxBlockGroup::SetBlockDuration(uint64 TimeLength)
 {
   assert(ParentTrack != nullptr);
-  int64 scale = ParentTrack->GlobalTimecodeScale();
-  KaxBlockDuration & myDuration = *static_cast<KaxBlockDuration *>(FindFirstElt(EBML_INFO(KaxBlockDuration), true));
+  const int64 scale = ParentTrack->GlobalTimecodeScale();
+  auto myDuration = *static_cast<KaxBlockDuration *>(FindFirstElt(EBML_INFO(KaxBlockDuration), true));
```
Comment (mbunkus) [inline src/KaxBlock.cpp]: This creates a _copy_ of the element, leading to invalid code execution. Please re-add the reference.

This bad change is modified further in the next commit, "remove unnecessary references". That chunk must be adjusted accordingly.

**Review by mbunkus — CHANGES_REQUESTED** [review]
With the fixes to all four places the MKVToolNix test suit finally runs without issues.

File: src/KaxBlock.cpp (hunk 4 above)
Comment (neheb) [reply src/KaxBlock.cpp]: Fixed.

File: src/KaxCues.cpp (hunk 3 above)
Comment (neheb) [reply src/KaxCues.cpp]: Fixed.

File: src/KaxCues.cpp (hunk 2 above)
Comment (neheb) [reply src/KaxCues.cpp]: Fixed.

File: src/KaxBlock.cpp (hunk 1 above)
Comment (neheb) [reply src/KaxBlock.cpp]: Got rid of the commit.

Comment (mbunkus) [reply src/KaxBlock.cpp]: The other hunks of that commit were fine; only this one hunk was problematic.

Even though the other hunks also used a `size_t` with `myBuffers.size() - 1`, it couldn't actually underflow as there's a check for `myBuffers.size() <= 1` with a subsequent `return` before those `- 1` calculations.

**Comment by mbunkus** [comment]
Thanks, this is fine now. Merged.

**Comment by mbunkus** [comment]
True. It did initially, though not anymore.

**Pull request merged**